physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,684 @@
+func (b *RaftBackend) SetAutopilotConfig(config *AutopilotConfig) {
+	b.l.Lock()
+	b.logger.Info("updating autopilot configuration", "config", config)
+	b.autopilotConfig = config
+	b.l.Unlock()
+}
+
+// AutopilotConfig returns the autopilot configuration in the backend.
+func (b *RaftBackend) AutopilotConfig() *AutopilotConfig {
+	b.l.RLock()
+	conf := b.autopilotConfig
+	b.l.RUnlock()
+	return conf.Clone()
+}
+
+func (b *RaftBackend) defaultAutopilotConfig() *AutopilotConfig {
+	return &AutopilotConfig{
+		CleanupDeadServers:             false,
+		LastContactThreshold:           10 * time.Second,
+		DeadServerLastContactThreshold: 24 * time.Hour,
+		MaxTrailingLogs:                1000,
+		MinQuorum:                      3,
+		ServerStabilizationTime:        10 * time.Second,
+	}
+}
+
+func (b *RaftBackend) AutopilotDisabled() bool {
+	b.l.RLock()
+	disabled := b.disableAutopilot
+	b.l.RUnlock()
+	return disabled
+}
+
+// StartAutopilot puts autopilot subsystem to work. This should only be called
+// on the active node.
+func (b *RaftBackend) StartAutopilot(ctx context.Context) {
+	b.l.RLock()
+	defer b.l.RUnlock()
+
+	if b.autopilot == nil {
+		return
+	}
+
+	b.autopilot.Start(ctx)
+	b.followerHeartbeatTrackerStopCh = make(chan struct{})
+	go b.startFollowerHeartbeatTracker()
+}
+
+// AutopilotExecutionStatus represents the current status of the autopilot background go routines
+type AutopilotExecutionStatus string
+
+const (
+	AutopilotNotRunning   AutopilotExecutionStatus = "not-running"
+	AutopilotRunning      AutopilotExecutionStatus = "running"
+	AutopilotShuttingDown AutopilotExecutionStatus = "shutting-down"
+)
+
+func autopilotStatusToStatus(status autopilot.ExecutionStatus) AutopilotExecutionStatus {
+	switch status {
+	case autopilot.Running:
+		return AutopilotRunning
+	case autopilot.ShuttingDown:
+		return AutopilotShuttingDown
+	default:
+		return AutopilotNotRunning
+	}
+}
+
+func (b *RaftBackend) startFollowerHeartbeatTracker() {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-b.followerHeartbeatTrackerStopCh:
+			return
+		case <-ticker.C:
+			markingCandidatePresent := false
+			b.l.RLock()
+			if b.autopilotConfig.CleanupDeadServers && b.autopilotConfig.DeadServerLastContactThreshold != 0 {
+				b.followerStates.l.RLock()
+				for _, state := range b.followerStates.followers {
+					if state.LastHeartbeat.IsZero() || state.IsDead {
+						continue
+					}
+					now := time.Now()
+					if now.After(state.LastHeartbeat.Add(b.autopilotConfig.DeadServerLastContactThreshold)) {
+						markingCandidatePresent = true
+					}
+				}
+				b.followerStates.l.RUnlock()
+
+				// If marking a node as dead, switch read with a write lock and
+				// check the conditions again.
+				if markingCandidatePresent {
+					b.followerStates.l.Lock()
+					for id, state := range b.followerStates.followers {
+						if state.LastHeartbeat.IsZero() || state.IsDead {
+							continue
```

Comment:
Same comment here as with the above `continue` re: unlocking

Reply:
Same response as above :-)